### PR TITLE
Add the support for dedicated claim dialect

### PR DESCRIPTION
**Pull request by supunharsha**

### CODE REVIEW

File: components/org.wso2.carbon.identity.application.authenticator.facebook/src/main/java/org/wso2/carbon/identity/application/authenticator/facebook/FacebookAuthenticator.java [hunk 1]
```diff
@@ -337,19 +338,320 @@ protected void setSubject(AuthenticationContext context, Map<String, Object> jso
...
-            for (Map.Entry<String, Object> entry : jsonObject.entrySet()) {
-                claims.put(ClaimMapping.build(entry.getKey(), entry.getKey(), null,
-                        false), entry.getValue().toString());
-                if (log.isDebugEnabled() &&
-                        IdentityUtil.isTokenLoggable(IdentityConstants.IdentityTokens.USER_CLAIMS)) {
-                    log.debug("Adding claim mapping : " + entry.getKey() + " <> " + entry.getKey() + " : "
-                            + entry.getValue());
+            Map<ClaimMapping, String> claims = new HashMap<>();
+            String claimUri;
+            String userInfoValue;
```
Comment (supunharsha) [inline components/org.wso2.carbon.identity.application.authenticator.facebook/src/main/java/org/wso2/carbon/identity/application/authenticator/facebook/FacebookAuthenticator.java]: claimUri

File: components/org.wso2.carbon.identity.application.authenticator.facebook/src/main/java/org/wso2/carbon/identity/application/authenticator/facebook/FacebookAuthenticator.java [hunk 2]
```diff
@@ -337,19 +338,320 @@ protected void setSubject(AuthenticationContext context, Map<String, Object> jso
...
+            String userInfoKey;
```
Comment (supunharsha) [inline components/org.wso2.carbon.identity.application.authenticator.facebook/src/main/java/org/wso2/carbon/identity/application/authenticator/facebook/FacebookAuthenticator.java]: claimValue

File: components/org.wso2.carbon.identity.application.authenticator.facebook/src/main/java/org/wso2/carbon/identity/application/authenticator/facebook/FacebookAuthenticator.java [hunk 3]
```diff
@@ -337,19 +338,320 @@ protected void setSubject(AuthenticationContext context, Map<String, Object> jso
...
+                            } else {
+                                if (log.isErrorEnabled()) {
+                                    log.error("The recieved user information from facebook cannot be mapped to any " +
+                                            "user information constants");
+                                }
+                                throw new ApplicationAuthenticatorException("User claim building failure");
+                            }
+                    }
+                } else {
+                    if (log.isWarnEnabled()) {
```
Comment (supunharsha) [inline components/org.wso2.carbon.identity.application.authenticator.facebook/src/main/java/org/wso2/carbon/identity/application/authenticator/facebook/FacebookAuthenticator.java]: Add DEBUG

File: components/org.wso2.carbon.identity.application.authenticator.facebook/src/main/java/org/wso2/carbon/identity/application/authenticator/facebook/FacebookAuthenticator.java [hunk 4]
```diff
@@ -337,19 +338,320 @@ protected void setSubject(AuthenticationContext context, Map<String, Object> jso
...
+                        log.warn("The key or/and value of user information came from facebook is null or empty");
```
Comment (darshanasbg) [inline components/org.wso2.carbon.identity.application.authenticator.facebook/src/main/java/org/wso2/carbon/identity/application/authenticator/facebook/FacebookAuthenticator.java]: use debug level to log this info and add username

File: components/org.wso2.carbon.identity.application.authenticator.facebook/src/main/java/org/wso2/carbon/identity/application/authenticator/facebook/FacebookAuthenticator.java (hunk 1 above)
Comment (supunharsha) [reply components/org.wso2.carbon.identity.application.authenticator.facebook/src/main/java/org/wso2/carbon/identity/application/authenticator/facebook/FacebookAuthenticator.java]: done

File: components/org.wso2.carbon.identity.application.authenticator.facebook/src/main/java/org/wso2/carbon/identity/application/authenticator/facebook/FacebookAuthenticator.java (hunk 2 above)
Comment (supunharsha) [reply components/org.wso2.carbon.identity.application.authenticator.facebook/src/main/java/org/wso2/carbon/identity/application/authenticator/facebook/FacebookAuthenticator.java]: done

**Pull request closed without merging**